cpp/ql/src/semmle/code/cpp/Type.qll

```diff
@@ -144,6 +145,66 @@ class Type extends Locatable, @type {
    */
   string explain() { result = "type" } // Concrete base impl to allow filters on Type
 
+  /**
+   * Gets a string that uniquely identifies this type, suitable for use when debugging queries. All typedefs and
+   * decltypes are expanded, and all symbol names are fully qualified.
+   *
+   * This operation is very expensive, and should not be used in production queries.
+   */
+  final string getTypeIdentityString() {
+    /*
+      The identity string of a type is just the concatenation of the four components below. To create the type
+      identity for a derived type, insert the declarator of the derived type between the `getDeclaratorPrefix()` and
+      `getDeclaratorSuffixBeforeQualifiers()`. To create the type identity for a `SpecifiedType`, insert the qualifiers
+      after `getDeclaratorSuffixBeforeQualifiers()`.
+    */
+    result = getTypeSpecifier() + getDeclaratorPrefix() + getDeclaratorSuffixBeforeQualifiers() + getDeclaratorSuffix()
+  }
+
+  /**
+   * Gets the "type specifier" part of this type's name. This is generally the "leaf" type from which the type was
+   * constructed.
+   *
+   * Examples:
+   * - `int` -> `int`
+   * - `int*` -> `int`
+   * - `int (*&)(float, double) const` -> `int`
+   *
+   * This predicate is intended to be used only by the implementation of `getTypeIdentityString`.
+   */
+  string getTypeSpecifier() {
+    result = ""
+  }
+
+  /**
+   * Gets the portion of this type's declarator that comes before the declarator for any derived type.
+   *
+   * This predicate is intended to be used only by the implementation of `getTypeIdentityString`.
+   */
+  string getDeclaratorPrefix() {
+    result = ""
+  }
+
+  /**
+   * Gets the portion of this type's declarator that comes after the declarator for any derived type, but before any
+   * qualifiers on the current type.
+   *
+   * This predicate is intended to be used only by the implementation of `getTypeIdentityString`.
+   */
+  string getDeclaratorSuffixBeforeQualifiers() {
+    result = ""
+  }
+
+  /**
+   * Gets the portion of this type's declarator that comes after the declarator for any derived type and after any
+   * qualifiers on the current type.
+   *
+   * This predicate is intended to be used only by the implementation of `getTypeIdentityString`.
+   */
+  string getDeclaratorSuffix() {
+    result = ""
+  }
```

Comment:
These four helper predicates are too likely to cause confusion when users browse the autocomplete list. I know it's convenient to use OO like this to ensure all cases are covered, but we don't even get that benefit as long as we can't use abstract predicates.

Reply:
This isn't the first time that I've wished that QL had the `protected` access specifier.
I've moved these predicates into private classes in the `Print` module.